plugins/zenoh-plugin-trait/src/plugin.rs

```diff
@@ -0,0 +1,240 @@
+    /// This can be overridden by the plugin implementation if the plugin is able to report its status: no connection to the database, etc.
+    fn report(&self) -> PluginReport {
+        PluginReport::default()
+    }
+    /// Collects information of sub-plugins matching the `_names` key expression. The information is richer than the one returned by `report()`: it contains external information about the running plugin, such as its name, path on disk, load status, etc.
+    /// Returns an empty list by default.
+    fn plugins_status(&self, _names: &keyexpr) -> Vec<PluginStatusRec> {
+        Vec::new()
+    }
+}
+
+pub trait PluginStartArgs: StructVersion {}
+
+pub trait PluginInstance: StructVersion + PluginControl + Send {}
+
+/// Base plugin trait. The loaded plugin
+pub trait Plugin: Sized + 'static {
+    type StartArgs: PluginStartArgs;
+    type Instance: PluginInstance;
+    /// Plugins' default name when statically linked.
+    const DEFAULT_NAME: &'static str;
+    /// Plugin's version. Used only for information purposes. It's recommended to use [plugin_version!] macro to generate this string.
+    const PLUGIN_VERSION: &'static str;
+    /// Plugin's long version (with git commit hash). Used only for information purposes. It's recommended to use [plugin_long_version!] macro to generate this string.
+    const PLUGIN_LONG_VERSION: &'static str;
+    /// Starts your plugin. Use `Ok` to return your plugin's control structure
+    fn start(name: &str, args: &Self::StartArgs) -> ZResult<Self::Instance>;
+}
+
+#[macro_export]
+macro_rules! plugin_version {
+    () => {
+        env!("CARGO_PKG_VERSION")
+    };
+}
+
+#[macro_export]
+macro_rules! plugin_long_version {
+    () => {
+        git_version::git_version!(prefix = "v", cargo_prefix = "v")
+    };
+}
+
+impl PluginReport {
+    pub fn new() -> Self {
+        Self::default()
+    }
+    pub fn clear(&mut self) {
+        *self = Self::default();
+    }
+    pub fn get_level(&self) -> PluginReportLevel {
+        self.level
+    }
+    pub fn add_error<S: Into<Cow<'static, str>>>(&mut self, error: S) {
+        self.level |= PluginReportLevel::Error;
+        self.messages.push(error.into());
+    }
+    pub fn add_warning<S: Into<Cow<'static, str>>>(&mut self, warning: S) {
+        self.level |= PluginReportLevel::Warning;
+        self.messages.push(warning.into());
+    }
+    pub fn add_message<S: Into<Cow<'static, str>>>(&mut self, message: S) {
+        self.level |= PluginReportLevel::Normal;
+        self.messages.push(message.into());
+    }
+    pub fn messages(&self) -> &[Cow<'static, str>] {
+        &self.messages
+    }
+}
+
+pub trait PluginConditionSetter {
+    fn add_error(self, report: &mut PluginReport) -> Self;
+    fn add_warning(self, report: &mut PluginReport) -> Self;
+    fn add_message(self, report: &mut PluginReport) -> Self;
```

Comment:
As log levels we have "info, warn, err ,debug, trace". Do `add_error`, `add_warning`, and `add_message` refer to those levels? If so, would `add_message` better be named as `add_info`?

Reply:
No, this is related to plugin's error report level:
```
pub enum PluginReportLevel {
    #[default]
    Normal,
    Warning,
    Error,
}
```
I'll think about giving names which shows this relation more clear.

Reply:
Ok, maybe `Normal` could be renamed to `Info` then? What do you think?

Reply:
Agree. This enum was "PluginStatus" before, the "Normal" remained from this. Now "Info" is ok here.